python/pfs/utils/optimizeBlackDots.py

```diff
@@ -0,0 +1,535 @@
+    def check(x, y, xc, yc, r=None):
+        """Check if the cobra is covered by the black dot
+
+        Parameters
+        ----------
+        x: `float`
+            x coordinates of the dot
+        y: `float`
+            y coordinate of the dot
+        xc: `float`
+            x coordinate of the cobra
+        yc: `float`
+            y coordinate of the cobra
+        r: `float`
+            radius of dot [mm]
+
+        Returns
+        ----------
+        cover value: `int`
+            returns 1 if the cobra is covered, 0 if not
+
+        Notes
+        ----------
+        Gets called by `quality_measure_single_spot`
+        """
+        if r is None:
+            r = 0.75
+        return int(((x - xc)**2 + (y - yc)**2) < r**2)
+
+    def new_position_of_dot(self, i, xd_mod, yd_mod, scale, rot, x_scale_rot, y_scale_rot):
+        """Move the dot to the new position
+
+        Parameters
+        ----------
+        i: `integer`
+            index of the cobra, starting from 1
+        xd_mod: `float`
+            x offset of the new position
+        yd_mod: `float`
+            y offset of the new position
+        scale: `float`
+            overall scaling
+        x_scale_rot: `float`
+            x coordinate around which to rotate
+        x_scale_rot: `float`
+            y coordinate around which to rotate
+
+        Returns
+        ----------
+        xd_new, yd_new: `float`, `float`
+            x and y coordinate of the new position
+
+        Notes
+        ---------
+        This function allows the user to move the dots in x and y direction,
+        and rotate/scale around custom origin
+
+        Calls `rotate`
+        """
+        dots = self.dots
+
+        xd_original = dots.loc[i-1]['x']
+        yd_original = dots.loc[i-1]['y']
+
+        xd_offset = xd_original - x_scale_rot
+        yd_offset = yd_original - y_scale_rot
+
+        r_distance_from_center = np.sqrt(xd_offset**2 + yd_offset**2)
+
+        xc_rotated, yc_rotated = self.rotate((xd_offset*(1+scale*r_distance_from_center/100),
+                                              yd_offset*(1+scale*r_distance_from_center/100)), (0, 0), rot)
+
+        xc_rotated = xc_rotated+x_scale_rot
+        yc_rotated = yc_rotated+y_scale_rot
+
+        xd_new = xc_rotated + xd_mod
+        yd_new = yc_rotated + yd_mod
+        return xd_new, yd_new
+
+    def quality_measure_single_spot(self, list_of_prepared_observations_and_predictions_single_fiber,
+                                    xd, yd):
+        """Calculate a penalty given x and y dot positions
+
+        Parameters
+        ----------
+        list_of_prepared_observations_and_predictions: `list`
+            list contaning observations for only one fiber
+        xd: `float`
+            x position of the dot
+        yd: `float`
+            y position of the dot
+
+        Returns
+        ----------
+        quality_measure: `float`
+            total penalty for this position of the dot
```

Comment:
Is it a quality (higher number is better) or a penalty (lower number is better)?